### ISSUE TITLE
Multiple epochs for Incremental

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask-ml#264]
Incremental should have a `max_iter` hyperparameter to control the number of passes (epochs) over the training data.

Implementing a `max_iter` is relatively straightforward. What we really want is some kind of "as many epochs as necessary until the underlying estimator has converged" parameter, which seems relatively difficult.

Split from https://github.com/dask/dask-ml/pull/258

**Comment by mrocklin** [comment]
@ogrisel suggested that it might make sense to score frequently and then have some sort of patience parameter that tells us to stop training after our score has not improved on a validation set.

**Comment by TomAugspurger** [comment]
Yes, he mentioned that to me as well. Thanks for capturing it.

> not improved on a validation set

Should this validation set be provided by the user, or should `X` and `y` be split internal to `Incremental`?

**Comment by mrocklin** [comment]
I imagine at least optionally provided by the user.  I don't know what we
should do in the case where they don't provide anything.

**Comment by ogrisel** [comment]
The `partial_fit` will not provide convergence checks by it-self. However, it would be great to replicate the convergence checks done in:

https://github.com/scikit-learn/scikit-learn/pull/9043

That is, set `n_iter_no_change=5`, `tol=1e-3`, `max_iter=1000, validation_split=0.1` as default values in `Incremental`.

If `validation_split=None`, then disable the convergence check and ensure you reach `max_iter` for instance.

**Comment by ogrisel** [comment]
Or alternatively, one could also use a subsample of the training set if not `validation_split` is passed. In scikit-learn we have an `early_stopping=True/False` flag to switch from validation vs training set based convergence check but I am not a big fan of the name of that switch.

**Comment by ogrisel** [comment]
Making it possible for the user to explicitly pass a pair of `(X, y)` dask arrays (or just a single `X` dask array for unsupervised models such as Minibatch K-Means) is also an option. Maybe this should be a `validation_data=(X, y)` fit-params rather than a constructor param in that case.

**Comment by mrocklin** [comment]
For stopping conditions it seems like there are two options:

1.  Use dask.delayed with a large graph of size `max_iter` and pass through a sentinel value if we should complete early.  We'll still run all of the tasks, but the latter part of them will run very quickly.

    This is nice because it uses only the local scheduler but introduces some overhead with very many tasks.

2.  Use dask.distributed and the concurrent.futures interface.  We would submit a few extra tasks to lead the computation, but would collect scores as they complete and not start any more computation if the stopping condition is met.

    This is nice because it is more efficient and dynamic, but introduces a requirement on the distributed scheduler

**Comment by ogrisel** [comment]
Do you think it would be a huge effort to augment the local thread/process based scheduler with the `Client` API of the distributed scheduler?

**Comment by mrocklin** [comment]
It would be moderate effort.  Probably a 2-3 weeks from someone who has
worked on the local scheduler before.

If that's the route that we go then I would suggest that we implement this
with the distributed scheduler today and then hope that the work above
happens in the future.

Also, client = Client(processes=False) is pretty lightweight.  We're adding
tornado and msgpack, but not much else.

**Comment by mrocklin** [comment]
If we're using the concurrent.futures API then I'm inclined to implement this with common machinery for RandomSearchCV and HyperBand.  See comment in https://github.com/dask/dask-ml/pull/288#issuecomment-405400091
